beacon_node/store/src/chunked_vector.rs

```diff
@@ -0,0 +1,611 @@
+    let (start_vindex, end_vindex) = F::start_and_end_vindex(slot, spec);
+    let start_cindex = start_vindex / chunk_size;
+    let end_cindex = end_vindex / chunk_size;
+
+    let chunks: Vec<Chunk<F::Value>> = range_query(store, F::column(), start_cindex, end_cindex)?;
+
+    let mut result = Vec::with_capacity(chunk_size * chunks.len());
+
+    for (chunk_index, chunk) in chunks.into_iter().enumerate() {
+        for (i, value) in chunk.values.into_iter().enumerate() {
+            let vindex = chunk_index * chunk_size + i;
+
+            if vindex >= start_vindex && vindex < end_vindex {
+                result.push(value);
+            }
+        }
+    }
+
+    Ok(result.into())
+}
+
+/// A chunk of a fixed-size vector from the `BeaconState`, stored in the database.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Chunk<T> {
+    /// A vector of up-to `chunk_size` values.
+    pub values: Vec<T>,
+}
+
+impl<T> Default for Chunk<T>
+where
+    T: Decode + Encode,
+{
+    fn default() -> Self {
+        Chunk { values: vec![] }
+    }
+}
+
+impl<T> Chunk<T>
+where
+    T: Decode + Encode,
+{
+    pub fn new(values: Vec<T>) -> Self {
+        Chunk { values }
+    }
+
+    pub fn load<S: Store>(store: &S, column: DBColumn, key: &[u8]) -> Result<Option<Self>, Error> {
+        store
+            .get_bytes(column.into(), key)?
+            .map(|bytes| Self::decode(&bytes))
+            .transpose()
+    }
+
+    pub fn store<S: Store>(&self, store: &S, column: DBColumn, key: &[u8]) -> Result<(), Error> {
+        store.put_bytes(column.into(), key, &self.encode()?)?;
+        Ok(())
+    }
+
+    /// Attempt to decode a single chunk, returning the chunk and the number of bytes read.
+    pub fn decode(bytes: &[u8]) -> Result<Self, Error> {
+        // NOTE: could have a sub-trait for fixed length SSZ types?
+        if !<T as Decode>::is_ssz_fixed_len() {
+            return Err(Error::from(ChunkError::InvalidType));
+        }
+
+        // Read the appropriate number of values
+        let mut offset = 0;
+        let mut values = vec![];
+        let value_size = <T as Decode>::ssz_fixed_len();
+
+        while offset < bytes.len() {
+            let value_bytes =
+                bytes
+                    .get(offset..offset + value_size)
+                    .ok_or(ChunkError::OutOfBounds {
+                        i: offset + value_size - 1,
+                        len: bytes.len(),
+                    })?;
+            let value = T::from_ssz_bytes(value_bytes)?;
+            values.push(value);
+            offset += value_size;
+        }
+
+        Ok(Chunk { values })
+    }
+
+    pub fn encoded_size(&self) -> usize {
+        self.values.len() * <T as Encode>::ssz_fixed_len()
+    }
+
+    /// Encode a single chunk as bytes.
+    pub fn encode(&self) -> Result<Vec<u8>, Error> {
+        // NOTE: could have a sub-trait for fixed length SSZ types?
+        if !<T as Decode>::is_ssz_fixed_len() {
```

Comment:
I can't see any reason why the `Encode` and `Decode` traits should differ here, but perhaps we should use `Encode`?